discovery_engine_core/core/src/storage/sqlite/utils.rs

```diff
@@ -39,3 +43,75 @@ where
         self
     }
 }
+
+pub(crate) trait SqlxSqliteErrorExt<V> {
+    /// Use this on the result of a sqlite query which might have failed a foreign
+    /// key constraint when an illegal document was passed in.
+    ///
+    /// For example it can be used in `.user_reacted` to handle it being
+    /// called with a random document id.
+    ///
+    /// If there is an error and it is a fk violation an appropriate error variant is
+    /// used instead of the default generic database error.
+    fn fk_violation_is_invalid_document_id(self, id: document::Id) -> Result<V, Error>;
+
+    fn on_row_not_found(self, error: Error) -> Result<V, Error>;
+}
+
+impl<V> SqlxSqliteErrorExt<V> for Result<V, sqlx::Error> {
+    fn fk_violation_is_invalid_document_id(self, id: document::Id) -> Result<V, Error> {
+        if let Err(sqlx::Error::Database(db_err)) = &self {
+            if db_err.code() == Some(Cow::Borrowed("787")) {
+                return Err(Error::InvalidDocumentId(id));
+            }
+        }
+        self.map_err(Into::into)
+    }
+
+    fn on_row_not_found(self, error: Error) -> Result<V, Error> {
+        if let Err(sqlx::Error::RowNotFound) = &self {
+            Err(error)
+        } else {
+            self.map_err(Into::into)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::storage::sqlite::SqliteStorage;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_fk_violation_is_invalid_document() {
+        let storage = SqliteStorage::connect("sqlite::memory:").await.unwrap();
+
+        sqlx::query("CREATE TABLE Foo(x INTEGER PRIMARY KEY);")
+            .execute(&storage.pool)
+            .await
+            .unwrap();
+
+        sqlx::query("CREATE TABLE Bar(x INTEGER PRIMARY KEY REFERENCES Foo(x));")
+            .execute(&storage.pool)
+            .await
+            .unwrap();
+
+        let document_id = document::Id::new();
+
+        let res = sqlx::query("INSERT INTO Bar(x) VALUES (?);")
+            .bind(10)
```

Comment:
❓ not sure if i understand this correctly, shouldn't this bind the `document_id`?

Reply:
In sqlite who only know that a FK-violation happened but not which fields it touched.

Wrt. this test I just test what the method checks, out of context of everything else. While I could rename `x` to `document_id` it could lead to some misunderstanding of what the method does.

This is a bit less misleading I hope  with how I changed the method in context of this review.

As a side note: This is different in postgres where you can get additional information and can name the check/fk/unique constraints. So in case of postgres we could have some more reliable/automatic conversion. This is (one of) the reasons why this extension is sqlite specific (and why I called it `SqlxSqliteErrorExt`).